staff/staff_views_helper.py

```diff
@@ -439,13 +513,23 @@ def get_course_attendance_statistic(course_id):
 
 def get_my_announcements(request, group):
     classroom = get_classroom_by_django_user(request.user)
-    announcements = (Announcement.objects.filter(recipient_groups=Group.objects.get(name=group)) | Announcement.objects.filter(recipient_classrooms=classroom)).distinct()
-    announcement_distributions = AnnouncementDistribution.objects.filter(announcement__in=announcements, user_id=request.user.id, dismissed=False)
+    announcements = (
+        Announcement.objects.filter(recipient_groups=Group.objects.get(name=group))
+        | Announcement.objects.filter(recipient_classrooms=classroom)
+    ).distinct()
```

Comment:
I think it is necessary here, but a little confused now, see images:

<img width="681" alt="Screen Shot 2019-08-15 at 11 12 03 AM" src="https://user-images.githubusercontent.com/37964808/63116166-c6c99f00-bf4d-11e9-967c-c46f5eb9792d.png">

<img width="263" alt="Screen Shot 2019-08-15 at 11 11 47 AM" src="https://user-images.githubusercontent.com/37964808/63116185-d2b56100-bf4d-11e9-9bc0-959b00def404.png">

Reply:
The part you're confused about is that you should be looking at is the length of the result (`announcement_distributions` and `form_distributions`), those will be the same. The actual inputs (`announcements`, `forms`) shouldn't be queried directly by django at all, it should just go into building the SQL query to compute the final result. You should be able to see difference if you turn on SQL query logging. https://docs.djangoproject.com/en/2.2/topics/logging/#django-db-backends

`len(forms)` is preferred to `forms.__len__()`, the latter is more of an implementation detail.

Reply:
Should probably also change `Group.objects.get(name=group)` to `Group.objects.filter(name=group)` to avoid an unnecessary query

Reply:
for example, this QuerySet is done in one query to the database:

```python
>>> AnnouncementDistribution.objects.filter(
    announcement__in=Announcement.objects.filter(
        recipient_groups__in=Group.objects.filter(name='staff')
    )
)
(0.002) SELECT "home_announcementdistribution"."id", "home_announcementdistribution"."announcement_id", "home_announcementdistribution"."user_id", "home_announcementdistribution"."dismissed" FROM "home_announcementdistribution" WHERE "home_announcementdistribution"."announcement_id" IN (SELECT V0."id" FROM "home_announcement" V0 INNER JOIN "home_announcement_recipient_groups" V1 ON (V0."id" = V1."announcement_id") WHERE V1."group_id" IN (SELECT U0."id" FROM "auth_group" U0 WHERE U0."name" = 'staff'))  LIMIT 21; args=('staff',)
<QuerySet []>
```

Reply:
To turn on query logging you can add this to settings.py:

```python
LOGGING = {
    'version': 1,
    'filters': {
        'require_debug_true': {
            '()': 'django.utils.log.RequireDebugTrue',
        }
    },
    'handlers': {
        'console': {
            'level': 'DEBUG',
            'filters': ['require_debug_true'],
            'class': 'logging.StreamHandler',
        }
    },
    'loggers': {
        'django.db.backends': {
            'level': 'DEBUG',
            'handlers': ['console']
        }
    }
}
```

Reply:
> The part you're confused about is that you should be looking at is the length of the result (`announcement_distributions` and `form_distributions`), those will be the same. The actual inputs (`announcements`, `forms`) shouldn't be queried directly by django at all, it should just go into building the SQL query to compute the final result. You should be able to see difference if you turn on SQL query logging. https://docs.djangoproject.com/en/2.2/topics/logging/#django-db-backends
> 
> `len(forms)` is preferred to `forms.__len__()`, the latter is more of an implementation detail.

Wow I had no idea about SQL query logging, this is great, I'm a bit overwhelmed by all the queries I'm seeing but it will be extremely useful!

Reply:
> for example, this QuerySet is done in one query to the database:
> 
> ```python
> >>> AnnouncementDistribution.objects.filter(
>     announcement__in=Announcement.objects.filter(
>         recipient_groups__in=Group.objects.filter(name='staff')
>     )
> )
> (0.002) SELECT "home_announcementdistribution"."id", "home_announcementdistribution"."announcement_id", "home_announcementdistribution"."user_id", "home_announcementdistribution"."dismissed" FROM "home_announcementdistribution" WHERE "home_announcementdistribution"."announcement_id" IN (SELECT V0."id" FROM "home_announcement" V0 INNER JOIN "home_announcement_recipient_groups" V1 ON (V0."id" = V1."announcement_id") WHERE V1."group_id" IN (SELECT U0."id" FROM "auth_group" U0 WHERE U0."name" = 'staff'))  LIMIT 21; args=('staff',)
> <QuerySet []>
> ```

Beautiful!

Reply:
You can also print the query property of a QuerySet at the console if you don't want to see every query get logged.

```python
>>> print(
    AnnouncementDistribution.objects.filter(
        announcement__in=Announcement.objects.filter(
            recipient_groups__in=Group.objects.filter(name='staff')
        )
    ).query
)
SELECT "home_announcementdistribution"."id", "home_announcementdistribution"."announcement_id", "home_announcementdistribution"."user_id", "home_announcementdistribution"."dismissed" FROM "home_announcementdistribution" WHERE "home_announcementdistribution"."announcement_id" IN (SELECT V0."id" FROM "home_announcement" V0 INNER JOIN "home_announcement_recipient_groups" V1 ON (V0."id" = V1."announcement_id") WHERE V1."group_id" IN (SELECT U0."id" FROM "auth_group" U0 WHERE U0."name" = staff))
```

Reply:
Reading through https://docs.djangoproject.com/en/2.2/ref/models/querysets/ is probably helpful

Reply:
> You can also print the query property of a QuerySet at the console if you don't want to see every query get logged.
> 
> ```python
> >>> print(
>     AnnouncementDistribution.objects.filter(
>         announcement__in=Announcement.objects.filter(
>             recipient_groups__in=Group.objects.filter(name='staff')
>         )
>     ).query
> )
> ```
> 
> ```sql
> SELECT "home_announcementdistribution"."id", "home_announcementdistribution"."announcement_id", "home_announcementdistribution"."user_id", "home_announcementdistribution"."dismissed" FROM "home_announcementdistribution" WHERE "home_announcementdistribution"."announcement_id" IN (SELECT V0."id" FROM "home_announcement" V0 INNER JOIN "home_announcement_recipient_groups" V1 ON (V0."id" = V1."announcement_id") WHERE V1."group_id" IN (SELECT U0."id" FROM "auth_group" U0 WHERE U0."name" = staff))
> ```

Excellent, yes this is easier for me to debug with, but from what I saw before it still looks like there are way, way, way too many queries when loading up a user's homepage.  I think this could be moved to an issue for the overall application backend functionality.

Reply:
I'd like to keep a copy of this conversation, is that possible?

Reply:
The conversations on github PRs don't go away when they are closed